python_modules/dagster/dagster/_model/__init__.py

```diff
@@ -58,3 +75,159 @@ def model_construct(cls, **kwargs: Any) -> Self:
             return super().model_construct(**kwargs)  # type: ignore
         else:
             return super().construct(**kwargs)
+
+
+T = TypeVar("T", bound=Type)
+
+
+def _banned(*args, **kwargs):
+    raise Exception("This method is not allowed on `@dagster_model`s.")
+
+
+def _nt_dict(nt) -> Mapping[str, Any]:
+    return {
+        key: value
+        for key, value in zip(nt._fields, nt.__hidden_iter__())
+        if key != ALT_CACHED_METHOD_CACHE_FIELD
+    }
+
+
+def __cached_method_safe_eq__(self, other):
+    if type(self) is not type(other):
+        return False
+    return _nt_dict(self) == _nt_dict(other)
+
+
+def __cached_method_safe_hash__(self):
+    return hash(
+        tuple(
+            v
+            for k, v in zip(self._fields, self.__hidden_iter__())
+            if k != ALT_CACHED_METHOD_CACHE_FIELD
+        )
+    )
+
+
+def _dagster_model_transform(
+    cls: T,
+    *,
+    enable_cached_method: bool,
+) -> T:
+    field_set = {
+        **cls.__annotations__,
+        **({ALT_CACHED_METHOD_CACHE_FIELD: Any} if enable_cached_method else {}),
+    }
+    base = NamedTuple(f"_{cls.__name__}", field_set.items())
+    orig_new = base.__new__
+    eval_ctx = EvalContext.capture_from_frame(3)
+    checks = {
+        name: build_check_call(ttype=ttype, name=name, eval_ctx=eval_ctx)
+        for name, ttype in cls.__annotations__.items()
+    }
+
+    def __checked_new__(cls, *args, **kwargs):
+        for key, fn in checks.items():
+            fn(kwargs[key])
+
+        cache_fields = {ALT_CACHED_METHOD_CACHE_FIELD: {}} if enable_cached_method else {}
+        return orig_new(cls, **kwargs, **cache_fields)
+
+    base.__new__ = __checked_new__
+
+    if enable_cached_method:
+        base.__eq__ = __cached_method_safe_eq__
+        base.__hash__ = __cached_method_safe_hash__
+
+    return type(
+        cls.__name__,
+        (cls, base),
+        {
+            "__iter__": _banned,
+            "__getitem__": _banned,
+            "__hidden_iter__": base.__iter__,
+        },
+    )  # type: ignore
+
+
+@overload
+def dagster_model(
+    cls: T,
+) -> T: ...
+
+
+@overload
+def dagster_model(
+    *,
+    enable_cached_method: bool = False,
+) -> Callable[[T], T]: ...
+
+
+@dataclass_transform(
+    kw_only_default=True,
+    frozen_default=True,
+)
+def dagster_model(
+    cls: Optional[T] = None,
+    *,
+    enable_cached_method: bool = False,
+) -> Union[T, Callable[[T], T]]:
```

Comment:
imo we should have a `with_new` bool here

Reply:
Also seems like at this point we should have our own version of `cached_property`, support it and `cached_method` and then have an `enable_cached` bool.

Reply:
and lastly i think `checked` should be a bool. I think we should be able to turn it off for perf if necessary.

Reply:
> imo we should have a with_new bool here

that would be ideal but the `dataclass_transform` bit is for static checks, ill see if i can get a bool w/ `@overload` to work other wise I think this is the only option

Reply:
Tried it but didnt work, then found this from https://peps.python.org/pep-0681/

> When applied to an overload, the dataclass_transform decorator still impacts all usage of the function.

so i think we need two decorators